clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -1056,6 +1056,25 @@ bool hasTrailingZeros(mlir::cir::ConstArrayAttr attr) {
           }));
 }
 
+static mlir::Attribute
+lowerDataMemberAttr(mlir::ModuleOp moduleOp, mlir::cir::DataMemberAttr attr,
+                    const mlir::TypeConverter &typeConverter) {
+  mlir::DataLayout layout{moduleOp};
+
+  uint64_t memberOffset;
+  if (attr.isNullPtr())
+    memberOffset = -1ull;
```

Comment:
This constant value for null pointers is actually ABI-specific. Should we move this to some ABI-specific part?

Reply:
Should we elevate `lowerDataMemberAttr` as a whole to LoweringPrepareCXXABI, so that the codes are centralized?

Reply:
> Should we elevate lowerDataMemberAttr as a whole to LoweringPrepareCXXABI, so that the codes are centralized?

Transforming data member pointers to member offsets during lowering prepare sounds reasonable at first, but the tricky part here is you also have to change every occurrences of `!cir.data_member` type to `!cir.int` type, which can be hard to implement during lowering prepare.

Reply:
Work that @sitio-couto is doing in https://github.com/llvm/clangir/pull/643 should provide us a way to ask those queries at LLVM lowering stage as well, since the information should be available at any time post CIRGen. I think it's fine to leave this as is and add a missing feature so that @sitio-couto can patch this when he lands his work. @sitio-couto any extra thoughts?

Reply:
@bcardosolopes sounds good to me. But the "tricky part" that @Lancern pointed out will still be an issue on the ABI lowering stage. AFAIK, there's no way for passes to iterate on specific types. We would either need to walk through every op or have a well-defined set of ops that may use the `DataMemberAttr`.

Reply:
This seems to be the same thing we discussed in https://github.com/llvm/clangir/pull/652/files#r1625515020. To be more clear, I'm not suggesting we do it in LoweringPrepare, what I'm suggesting is that since @sitio-couto ABI library is independent of CIRGen, it should also be usable from LowerToLLVM.cpp.

Reply:
The PR got merged and `MissingFeatures::targetCodeGenInfoGetNullPointer()` is ready now.

Reply:
@Lancern happy to land this once you add the assertion on the missing feature!

Reply:
> happy to land this once you add the assertion on the missing feature!

Added :)